### PR TITLE
Prevent non-updating tiles by disallowing an expiration time of 0

### PR DESCRIPTION
**Pull request by kkaefer**
In-depth explanation in https://github.com/mapbox/mapbox-gl-native/issues/10015#issuecomment-334286498

### CODE REVIEW

**Comment by kkaefer** [comment]
Open questions are:

* [ ] How do we test this? It's not an issue with our implementation of `OnlineFileSource` in particular, rather, it's an issue with the way we (ab)use the 0 value in `TileLoader<T>` in combination with `DefaultFileSource`'s skipping of cache lookups.
* [ ] What do we do about existing caches that already contain stuck tiles with bogus expiration timeouts? A value of 2065 isn't *wrong* per se, so disallowing them isn't a great strategy. We could add a migration that deletes all cache items that have an expiration value that is beyond e.g. 2,147,483,647, but given that we don't have indices on the expiration column, this could be an expensive upgrade operation for large databases.

**Comment by jfirebaugh** [comment]
Could we avoid using zero as a special value instead (e.g. via the "more long term solution" you describe in https://github.com/mapbox/mapbox-gl-native/issues/10015#issuecomment-334286498)? Avoiding that kind of special value is why we use an `optional` here in the first place.

**Comment by kkaefer** [comment]
Previously, we used the existence of a `prior*` field in the Resource object as an indication for whether we should consult the cache or not. However, this is prone to error, since a failed cache lookup won't set any prior fields. Therefore, we manually set `priorExpires` to 0. This in turn triggered another bug where generated wrong expiration timestamps when the server response we got was expired (or expired between sending and receiving).

This commit changes the flags so that we can now explicitly request `CacheOnly`/`NetworkOnly` (or `All`) loading methods, rather than the implicit `Optional`/`Required` naming scheme.

**Comment by kkaefer** [comment]
@jfirebaugh can you please re-review? 🙇

File: src/mbgl/tile/tile_loader_impl.hpp [hunk 1]
```diff
@@ -25,21 +25,21 @@ TileLoader<T>::TileLoader(T& tile_,
...
+        loadFromCache();
     } else {
         // When the FileSource doesn't support optional requests, we do nothing until the
         // data is definitely required.
-        if (necessity == Necessity::Required) {
-            loadRequired();
+        if (necessity == TileNecessity::Required) {
+            loadFromNetwork();
         } else {
             // We're using this field to check whether the pending request is optional or required.
```
Comment (jfirebaugh) [inline src/mbgl/tile/tile_loader_impl.hpp]: "to check whether the pending request is a cache request or network request"

File: src/mbgl/tile/tile_loader_impl.hpp [hunk 2]
```diff
@@ -48,44 +48,45 @@ template <typename T>
...
+    resource.loadingMethod = Resource::LoadingMethod::CacheOnly;
     request = fileSource.request(resource, [this](Response res) {
         request.reset();
 
-        tile.setTriedOptional();
+        tile.setTriedCache();
 
         if (res.error && res.error->reason == Response::Error::Reason::NotFound) {
             // When the optional request could not be satisfied, don't treat it as an error.
             // Instead, we make sure that the next request knows that there has been an optional
```
Comment (jfirebaugh) [inline src/mbgl/tile/tile_loader_impl.hpp]: Update this comment too.

**Pull request merged**